File: rename.go
// rename.go -- rename file names (Go version)

package main

import (
	"os"
	"fmt"
	"strings"
)

func usage() {
	fmt.Println("usage: rename current_string replacement file_pattern")
	os.Exit(1)
}

func main() {
	if len(os.Args) < 4 {
		usage()
	}
	currentString := os.Args[1]
	replaceString := os.Args[2]
	files := os.Args[3:]
	var new_file string
	var err error
	for _, file := range files {
		new_file = strings.ReplaceAll(file, currentString, replaceString)
		err = os.Rename(file, new_file)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
	}
}

File: rename.cpp
// rename.cpp -- rename files (C++ version)
#include <iostream>
#include <string>
#include <cstdio>
using namespace std;

void usage() {
    cout << "usage: rename old_string new_string file_pattern" << endl;
}

int main(int argc, char **argv) {
    if (argc < 4) {
        usage();
        return 1;
    }
    string currentString = argv[1];
    string replaceString = argv[2];
    int currentStringLength = currentString.size();
    int replaceStringLength = replaceString.size();
    size_t pos;
    for (int i = 3; i < argc; i++) {
        string currentName = argv[i];
        string newName = currentName;
        pos = newName.find(currentString);
        while(pos != string::npos) {
            newName.replace(pos, currentStringLength, replaceString);
            pos = newName.find(currentString, pos + replaceStringLength);
        }
        if (rename(currentName.c_str(), newName.c_str()) != 0) {
            string err = "Error renaming file: " + currentName;
            perror(err.c_str());
        }
    }
    return 0;
}
